Readme.md
# retext-directionality [![Build Status](https://img.shields.io/travis/wooorm/retext-directionality.svg?style=flat)](https://travis-ci.org/wooorm/retext-directionality) [![Coverage Status](https://img.shields.io/coveralls/wooorm/retext-directionality.svg?style=flat)](https://coveralls.io/r/wooorm/retext-directionality?branch=master)

[**retext**](https://github.com/wooorm/retext "Retext") directionality detection.

## Installation

[npm](https://docs.npmjs.com/cli/install):

```bash
$ npm install retext-directionality
```

[Component.js](https://github.com/componentjs/component):

```bash
$ component install wooorm/retext-directionality
```

[Bower](http://bower.io/#install-packages):

```bash
$ bower install retext-directionality
```

[Duo](http://duojs.org/#getting-started):

```javascript
var directionality = require('wooorm/retext-directionality');
```

## Usage

```javascript
var Retext = require('retext');
var visit = require('retext-visit');
var inspect = require('retext-inspect');
var directionality = require('retext-directionality');

var retext = new Retext()
    .use(inspect)
    .use(visit)
    .use(directionality);

retext.parse('A simple, anglais, بسيطة.', function (err, tree) {
    /* Log the tree, note the data attributes: */
    console.log(tree);
    /*
     * RootNode[1] [data={"direction":"neutral"}]
     * └─ ParagraphNode[1] [data={"direction":"neutral"}]
     *    └─ SentenceNode[10] [data={"direction":"neutral"}]
     *       ├─ WordNode[1] [data={"direction":"ltr"}]
     *       │  └─ TextNode: 'A' [data={"direction":"ltr"}]
     *       ├─ WhiteSpaceNode: ' ' [data={"direction":"neutral"}]
     *       ├─ WordNode[1] [data={"direction":"ltr"}]
     *       │  └─ TextNode: 'simple' [data={"direction":"ltr"}]
     *       ├─ PunctuationNode: ',' [data={"direction":"neutral"}]
     *       ├─ WhiteSpaceNode: ' ' [data={"direction":"neutral"}]
     *       ├─ WordNode[1] [data={"direction":"ltr"}]
     *       │  └─ TextNode: 'anglais' [data={"direction":"ltr"}]
     *       ├─ PunctuationNode: ',' [data={"direction":"neutral"}]
     *       ├─ WhiteSpaceNode: ' ' [data={"direction":"neutral"}]
     *       ├─ WordNode[1] [data={"direction":"rtl"}]
     *       │  └─ TextNode: 'بسيطة' [data={"direction":"rtl"}]
     *       └─ PunctuationNode: '.' [data={"direction":"neutral"}]
     */
});
```

## API

None, **retext-directionality** automatically detects the direction of each [`Text`](https://github.com/wooorm/textom#textomtextvalue-nlcsttext) (using **[wooorm/direction](https://github.com/wooorm/direction)**), either `"ltr"`, `"rtl"`, or `"neutral"`), and stores the direction in `node.data.direction`.

All parents (such as words, sentences, paragraphs, root) also receive a `direction` property (`parent.data.direction`): If every [`Text`](https://github.com/wooorm/textom#textomtextvalue-nlcsttext) inside a parent has the same (or neutral) direction, the parent has the same direction, otherwise, the parent has a neutral direction.

## Performance

On a MacBook Air, **retext** performs about 27% slower with **retext-directionality**.

```text
           retext w/o retext-directionality
  228 op/s » A paragraph (5 sentences, 100 words)
   25 op/s » A section (10 paragraphs, 50 sentences, 1,000 words)

           retext w/ retext-directionality
  166 op/s » A paragraph (5 sentences, 100 words)
   18 op/s » A section (10 paragraphs, 50 sentences, 1,000 words)
```

## License

[MIT](LICENSE) © [Titus Wormer](http://wooorm.com)


History.md

0.1.8 / 2015-01-18
==================

  * Update copyright notice in `LICENSE` to include 2015
  * Add link to whole license in `Readme.md`
  * Add Duo as an instalation method in `Readme.md`
  * Add links to installation methods in `Readme.md`
  * Refactor fences code blocks in `Readme.md`
  * Refactor to adhere to strict jsdoc style
  * Add jscs-jsdoc configuration to `.jscs.json`
  * Add jscs-jsdoc as a dev-dependency
  * Refactor npm scripts for changes in npm
  * Update eslint
  * Update retext

0.1.7 / 2014-12-09
==================

 * Add benchmark
 * Fix wording in `Readme.md`
 * Refactor npm script targets in `package.json`
 * Add link to personal website to copyright in `Readme.md`
 * Update eslint, retext

0.1.6 / 2014-11-21
==================

 * Add flat badges to `Readme.md`
 * Refactor `Readme.md`
 * Add retext-inspect to dev-dependecies
 * Refactor npm script targets in `package.json`
 * Fix incorrect executive rights on `test.js`
 * Update retext

0.1.5 / 2014-10-27
==================

 * Add `.eslintrc`
 * Refactor to disallow space after object keys
 * Refactor for changes in retext
 * Update eslint, mocha, retext, retext-content, retext-visit

0.1.4 / 2014-10-21
==================

 * Refactor .jscs.json
 * Refactor test for changes in retext
 * Update retext

0.1.3 / 2014-10-15
==================

 * Update tests for changes in retext-visit
 * Update direction, retext, retext-content, retext-visit
 * Update .gitignore, .npmignore, bower ignore
 * Move spec to test.js

0.1.2 / 2014-09-26
==================

 * Refactor Readme.md
 * Refactor API
 * Add npm deployment to travis
 * Remove `npm update` from travis
 * Remove browser test
 * Update .gitignore, .npmignore
 * Update Installation in docs
 * Add bower.json
 * Fix property order in component.json, package.json
 * Update copyright in Readme.md
 * Remove testling
 * Refactor spec
 * Update dependencies

0.1.1 / 2014-07-22
==================

 * Parents now receive updates when non-text nodes get inserted/removed
 * Added components to gitignore
 * Added retext as a peer dependency
 * Updated documentation for latest changes

0.1.0 / 2014-07-12
==================

 * Added changelog
 * Fixed spec for changes in 16e492a
 * Added functionality to detect direction on non-word nodes
 * Directions can no longer be `null`, just `"neutral"`
 * Removed unused code
 * Removed functionality to browserify unit tests by default
 * Updated eslint, istanbul, retext, retext-visit, direction

0.0.4 / 2014-06-19
==================

 * Unit tests can now run in the browser (spec/index.html)
 * Fixed a bug where everything in ./spec/ was tested
 * Split the lint task into lint-api, lint-test, and lint-style
 * Updated dependency versions of direction, jscs, retext, and retext-visit

0.0.3 / 2014-06-17
==================

 * Fix code styleguide for latest jscs updates
 * Updated dependency versions of direction, jscs, retext, and retext-visit
 * Update retext developer dependency to 0.0.10

0.0.2 / 2014-06-14
==================

 * Fixed some typos, added some more words about how the direction of parents works
 * Added an “install with component” guide
 * Testling badge & browser support
 * SVG badges & coverage status badge
 * Fixed some new-lines before EOF
 * Added Testling to package
 * Refactored travis-ci, added coveralls
 * API and unit tests are now ESLint compliant an use direction; moved retext-visit to developer dependencies
 * Removed Makefile, instead opting for just package.json
 * Removed complexity-report from developer dependencies
 * Replaced JSHint with ESLint
 * Add component.json
 * Update dependency versions of istanbul, retext, and retext-visit
 * Added direction as a dependency
 * `-` is now an allowed left-sticked operator (e.g., `-1`)

0.0.1 / 2014-06-05
==================



test.js
'use strict';

/*
 * Dependencies.
 */

var Retext,
    directionality,
    visit,
    inspect,
    content,
    assert;

Retext = require('retext');
directionality = require('./');
visit = require('retext-visit');
content = require('retext-content');
inspect = require('retext-inspect');
assert = require('assert');

/*
 * Fixtures.
 */

var sentence,
    leftToRightWords,
    rightToLeftWords;

/*
 * Base English sentence.
 */

sentence = 'A simple, english, sentence.';

/*
 * French, baguette!
 */

leftToRightWords = 'Un simple anglais phrase'.split(' ');

/*
 * Arabic.
 *
 * I'm no good at Aabic---but Google says this is the
 * translation of the example. For our purpose validity
 * however does not matter---we just need some
 * right-to-left words.
 */

rightToLeftWords = 'أ الجملة الانجليزية بسيطة'.split(' ');

/*
 * Retext
 */

var retext;

retext = new Retext()
    .use(inspect)
    .use(visit)
    .use(content)
    .use(directionality);

/*
 * Tests
 */

describe('directionality()', function () {
    it('should be a `function`', function () {
        assert(typeof directionality === 'function');
    });

    retext.parse(sentence, function (err, tree) {
        it('should not throw', function (done) {
            done(err);
        });

        it('should process each `WordNode`', function () {
            var index;

            tree.visit(tree.WORD_NODE, function (node) {
                assert(node.data.direction === 'ltr');
            });

            /*
             * Cover some optimizations: when a node's
             * direction does not change, there is no
             * need to re-process parents.
             */

            index = -1;

            tree.visit(tree.WORD_NODE, function (node) {
                node[0].fromString(leftToRightWords[++index]);

                assert(node.data.direction === 'ltr');
            });
        });

        it('should process each `Parent`', function () {
            tree.visit(tree.WORD_NODE, function (node) {
                var parent;

                parent = node.parent;

                while (parent) {
                    assert(parent.data.direction === 'ltr');

                    parent = parent.parent;
                }
            });
        });

        it('should set each `direction` to `null` when a `WordNode` (no ' +
            'longer?) has a value',
            function () {
                tree.visit(tree.WORD_NODE, function (node) {
                    /*
                     * We use the less secure `fromString`
                     * on a TextNode here to test for
                     * `changetext` events.
                     */

                    node[0].fromString('');

                    assert(node.data.direction === 'neutral');
                });
            }
        );

        it('should re-process each `WordNode` and `Parent`', function () {
            var index;

            /*
             * First set some left-to-rigth words, so we
             * can check for `neutral` changes.
             */

            index = -1;

            tree.visit(tree.WORD_NODE, function (node) {
                node.replaceContent(leftToRightWords[++index]);

                assert(node.data.direction === 'ltr');
            });

            index = -1;

            tree.visit(tree.WORD_NODE, function (node) {
                var parent,
                    shouldBeDirection;

                node.replaceContent(rightToLeftWords[++index]);

                assert(node.data.direction === 'rtl');

                /*
                 * When all words are `rtl`, the parent
                 * should reflext this instead of showing
                 * `neutral`.
                 */

                shouldBeDirection = index === 3 ? 'rtl' : 'neutral';

                parent = node.parent;

                while (parent) {
                    assert(parent.data.direction === shouldBeDirection);

                    parent = parent.parent;
                }
            });
        });

        it('should set `Parent`s `direction` to `neutral` when all its ' +
            '`WordNode`s are removed',
            function () {
                var parent;

                parent = tree.head.head;

                tree.visit(tree.TEXT_NODE, function (node) {
                    node.remove();
                });

                while (parent) {
                    assert(parent.data.direction === 'neutral');

                    parent = parent.parent;
                }
            }
        );
    });
});


spec/retext-directionality.spec.js
'use strict';

/**
 * Dependencies.
 */

var directionality,
    visit,
    content,
    Retext,
    assert;

directionality = require('..');
visit = require('retext-visit');
content = require('retext-content');
Retext = require('retext');
assert = require('assert');

/**
 * Fixtures.
 */

var sentence,
    leftToRightWords,
    rightToLeftWords;

/**
 * Base English sentence.
 */

sentence = 'A simple, english, sentence.';

/**
 * French, baguette!
 */

leftToRightWords = 'Un simple anglais phrase'.split(' ');

/**
 * Arabic.
 *
 * I'm no good at Aabic---but Google says this is the
 * translation of the example. For our purpose validity
 * however does not matter---we just need some
 * right-to-left words.
 */

rightToLeftWords = 'أ الجملة الانجليزية بسيطة'.split(' ');

/**
 * Retext
 */

var retext;

retext = new Retext()
    .use(visit)
    .use(content)
    .use(directionality);

/**
 * Tests
 */

describe('directionality()', function () {
    it('should be a `function`', function () {
        assert(typeof directionality === 'function');
    });

    it('should expose an `attach` method', function () {
        assert(typeof directionality.attach === 'function');
    });

    retext.parse(sentence, function (err, tree) {
        it('should not throw', function (done) {
            done(err);
        });

        it('should process each `WordNode`', function () {
            var index;

            tree.visitType(tree.WORD_NODE, function (node) {
                assert(node.data.direction === 'ltr');
            });

            /**
             * Cover some optimizations: when a node's
             * direction does not change, there is no
             * need to re-process parents.
             */

            index = -1;

            tree.visitType(tree.WORD_NODE, function (node) {
                node[0].fromString(leftToRightWords[++index]);

                assert(node.data.direction === 'ltr');
            });
        });

        it('should process each `Parent`', function () {
            tree.visitType(tree.WORD_NODE, function (node) {
                var parent;

                parent = node.parent;

                while (parent) {
                    assert(parent.data.direction === 'ltr');

                    parent = parent.parent;
                }
            });
        });

        it('should set each `direction` to `null` when a `WordNode` (no ' +
            'longer?) has a value',
            function () {
                tree.visitType(tree.WORD_NODE, function (node) {
                    /**
                     * We use the less secure `fromString`
                     * on a TextNode here to test for
                     * `changetext` events.
                     */

                    node[0].fromString('');

                    assert(node.data.direction === 'neutral');
                });
            }
        );

        it('should re-process each `WordNode` and `Parent`', function () {
            var index;

            /**
             * First set some left-to-rigth words, so we
             * can check for `neutral` changes.
             */

            index = -1;

            tree.visitType(tree.WORD_NODE, function (node) {
                node.replaceContent(leftToRightWords[++index]);

                assert(node.data.direction === 'ltr');
            });

            index = -1;

            tree.visitType(tree.WORD_NODE, function (node) {
                var parent,
                    shouldBeDirection;

                node.replaceContent(rightToLeftWords[++index]);

                assert(node.data.direction === 'rtl');

                /**
                 * When all words are `rtl`, the parent
                 * should reflext this instead of showing
                 * `neutral`.
                 */

                shouldBeDirection = index === 3 ? 'rtl' : 'neutral';

                parent = node.parent;

                while (parent) {
                    assert(parent.data.direction === shouldBeDirection);

                    parent = parent.parent;
                }
            });
        });

        it('should set `Parent`s `direction` to `neutral` when all its ' +
            '`WordNode`s are removed',
            function () {
                var parent;

                parent = tree.head.head;

                tree.visitType(tree.TEXT_NODE, function (node) {
                    if (node.parent.type === node.WORD_NODE) {
                        node.remove();
                    }
                });

                while (parent) {
                    assert(parent.data.direction === 'neutral');

                    parent = parent.parent;
                }
            }
        );
    });
});


index.js
/**
 * Dependencies.
 */

var Retext = require('wooorm/retext@0.5.0');
var directionality = require('wooorm/retext-directionality@0.1.8');
var dom = require('wooorm/retext-dom@0.3.1');
var visit = require('wooorm/retext-visit@0.2.5');

/**
 * Retext.
 */

var retext = new Retext()
    .use(dom)
    .use(visit)
    .use(directionality);

/**
 * DOM elements.
 */

var $input = document.getElementsByTagName('textarea')[0];
var $output = document.getElementsByTagName('div')[0];

/**
 * Events
 */

var tree;

function oninputchange() {
    if (tree) {
        tree.toDOMNode().parentNode.removeChild(tree.toDOMNode());
    }

    retext.parse($input.value, function (err, root) {
        if (err) throw err;

        tree = root;

        tree.visit(function (node) {
            if (!node.DOMTagName || !node.data.direction) {
                return
            }

            if (node.data.direction !== 'neutral') {
                node.toDOMNode().setAttribute('dir', node.data.direction);
            } else {
                node.toDOMNode().setAttribute('data-dir', node.data.direction);
            }
        });

        $output.appendChild(tree.toDOMNode());
    });
}

$input.addEventListener('input', oninputchange);

oninputchange();


benchmark.js
'use strict';

var Retext,
    directionality;

/*
 * Dependencies.
 */

Retext = require('retext');
directionality = require('./');

/*
 * Dependencies.
 */

var retext,
    retextWithDirectionality;

retext = new Retext();
retextWithDirectionality = new Retext().use(directionality);

/*
 * Test data: A (big?) article (w/ 100 paragraphs, 500
 * sentences, 10,000 words);
 *
 * Source:
 *   http://www.gutenberg.org/files/10745/10745-h/10745-h.htm
 */

var sentence,
    paragraph,
    section;

sentence = 'Where she had stood was clear, and she was gone since Sir ' +
    'Kay does not choose to assume my quarrel.';

paragraph = 'Thou art a churlish knight to so affront a lady ' +
    'he could not sit upon his horse any longer. ' +
    'For methinks something hath befallen my lord and that he ' +
    'then, after a while, he cried out in great voice. ' +
    'For that light in the sky lieth in the south ' +
    'then Queen Helen fell down in a swoon, and lay. ' +
    'Touch me not, for I am not mortal, but Fay ' +
    'so the Lady of the Lake vanished away, everything behind. ' +
    sentence;

section = paragraph + Array(10).join('\n\n' + paragraph);

suite('retext w/o retext-directionality', function () {
    bench('A paragraph (5 sentences, 100 words)', function (done) {
        retext.parse(paragraph, done);
    });

    bench('A section (10 paragraphs, 50 sentences, 1,000 words)',
        function (done) {
            retext.parse(section, done);
        }
    );
});

suite('retext w/ retext-directionality', function () {
    bench('A paragraph (5 sentences, 100 words)', function (done) {
        retextWithDirectionality.parse(paragraph, done);
    });

    bench('A section (10 paragraphs, 50 sentences, 1,000 words)',
        function (done) {
            retextWithDirectionality.parse(section, done);
        }
    );
});
